src/modules/julia.rs

```diff
@@ -0,0 +1,558 @@
+//! Pretend to update julia packages
+use async_trait::async_trait;
+use instant::Instant;
+use rand::prelude::*;
+use std::fmt::Display;
+use yansi::Paint;
+
+use crate::args::AppConfig;
+use crate::data::JULIA_PACKAGES_LIST;
+use crate::generators::gen_hex_string;
+use crate::io::{csleep, cursor_up, dprint, erase_line, newline};
+use crate::modules::Module;
+
+// Redefine a more efficient `crate::io::print` function that doesn't type one character at a time.
+async fn print<S: Into<String>>(s: S) {
+    #[cfg(target_arch = "wasm32")]
+    {
+        write_to_xterm(s.into().as_str());
+    }
+
+    #[cfg(not(target_arch = "wasm32"))]
+    {
+        use std::io::stdout;
+        use std::io::Write;
+        print!("{}", s.into());
+        stdout().flush().unwrap();
+    }
+}
+
+#[derive(Debug)]
+pub struct Package<'a> {
+    pub name: &'a str,
+    pub id: &'a str,
+    pub versions: Vec<&'a str>,
+}
+
+pub struct Julia;
+
+#[async_trait(?Send)]
+impl Module for Julia {
+    fn name(&self) -> &'static str {
+        "julia"
+    }
+
+    fn signature(&self) -> String {
+        "julia --threads auto --project .".to_string()
+    }
+
+    async fn run(&self, appconfig: &AppConfig) {
+        let mut rng = thread_rng();
+
+        // Choose `num_packages` packages and `num_artifacts` artifacts
+        // non-repeating and in random order
+        let num_packages = rng.gen_range(10..150);
+        let num_artifacts = rng.gen_range(1..10);
+        let all_packages: Vec<&Package> = JULIA_PACKAGES_LIST
+            .choose_multiple(&mut rng, num_packages + num_artifacts)
+            .collect();
+        let (packages, artifacts) = all_packages.split_at(num_packages);
+
+        // root project of the julia session
+        let project = if rng.gen::<f32>() < 0.3f32 {
+            "@v1.7"
```

Comment:
I don't really understand this default. What's the reasoning here?

Reply:
If you simply run `julia`, the session starts with the default home project/environment corresponding to the Julia version. These environments are stored in `~/.julia/environments/v1.*`. When you enter the package prompt by typing `]`, the prompt changes from `julia> ` to `(@v1.7) > `.

If you run `julia --project` inside a folder that contains a Julia `Project.toml`, the home project is set to this instead.

I wanted to simulate the chance of being either one, to randomize even more.

Reply:
I forgot to mention that this distinction between global environment and specific project has consequences later on, for instance in the function `async fn update_project_and_manifest(project: &str)` some paths are constructed differently:

```rust
    if let Some(project) = project.strip_prefix('@') {
        project_path = format!("~/.julia/environments/{}/Project.toml", project);
        manifest_path = format!("~/.julia/environments/{}/Manifest.toml", project);
    } else {
        project_path = format!("~/Documents/code/julia/projects/{}.jl/Project.toml", project);
        manifest_path = format!("~/Documents/code/julia/projects/{}.jl/Manifest.toml", project);
    }
```